### PR TITLE
Support for making a CMake build system from evoke

### PR DESCRIPTION
**Pull request by jonhanks**
This is my first stab at making evoke more usable from systems like clion and a stab at #6.

The target of this is not to use cmake as a build system per say, but
to make a cmake structure that IDEs and tooling that only understand
cmake can consume.

It properly creates dependencies using the discovery logic in evoke.
However for system install components it does not do a proper cmake
find_package/....  Instead it creates interface libraries to define
relationships.

This moves the CMakeLists.txt file that was shipped to CMakeLists.txt.tmpl
to denote that it is a template type file.

A new command line options is introduced '-cm' which when given creates
a set of CMakeList.txt files.

To facilitate the use of global options from the evoke.conf file the
toolsets are extended, providing an api which can take an command line
option string and break it out into a set of compile, include, and link
directives which should be used.  This is stored in a structure which
the CMakeLists building code uses to set global options (per target).

Updated the gitignore list to ignore idea/clion/cmake files.

### CODE REVIEW

**Comment by jonhanks** [comment]
I should add a sample of the generated files.  The top level CMakeLists.txt when run on evoke is:
<pre>
cmake_minimum_required(VERSION 3.12)
project(evoke)

# This is an autogenerated file, do not hand modify.
# This file was generated by evoke to give cmake only IDEs
# An idea about the project structure.

# System components
add_library(evoke_system_lib_boost_system INTERFACE)
target_link_libraries(evoke_system_lib_boost_system INTERFACE
    boost_system
)
add_library(evoke_system_lib_boost_filesystem INTERFACE)
target_link_libraries(evoke_system_lib_boost_filesystem INTERFACE
    boost_filesystem
    evoke_system_lib_boost_system)
add_library(evoke_system_lib_boost_process INTERFACE)

add_subdirectory(toolsets)
add_subdirectory(reporter)
add_subdirectory(project)
add_subdirectory(fw)
add_subdirectory(utils)
add_subdirectory(evoke)
</pre>
<hr/>

Then one of the sub directories comes across as follows, this highlights how options were taken from my evoke.conf file and passed to the cmake system (-std=c++17 -O3 -g3 -pthread -I/opt/boost-1.67.0/include -L/opt/boost-1.67.0/lib), these are not baked into the code but come by config file.
<pre>
# This is an autogenerated file, do not hand modify.
# This file was generated by evoke to give cmake only IDEs
# An idea about the project structure.

add_library(reporter STATIC
    src/ConsoleReporter.cpp
    src/SimpleReporter.cpp
    src/Reporter.cpp
    include/Reporter.h
)

target_compile_options(reporter PRIVATE -std=c++17 -O3 -g3 -pthread)
target_include_directories(reporter PRIVATE /opt/boost-1.67.0/include)
target_include_directories(reporter PUBLIC
    ${CMAKE_CURRENT_SOURCE_DIR}/include
)
target_link_libraries(reporter PRIVATE -lpthread -L/opt/boost-1.67.0/lib)
target_link_libraries(reporter PRIVATE
    project
)
</pre>

<hr/>
And a header only component looks like this
<pre>
# This is an autogenerated file, do not hand modify.
# This file was generated by evoke to give cmake only IDEs
# An idea about the project structure.

add_library(utils INTERFACE
)

target_compile_options(utils INTERFACE -std=c++17 -O3 -g3 -pthread)
target_include_directories(utils INTERFACE /opt/boost-1.67.0/include)
target_include_directories(utils INTERFACE
    ${CMAKE_CURRENT_SOURCE_DIR}/include
)
target_link_libraries(utils INTERFACE -lpthread -L/opt/boost-1.67.0/lib)
</pre>

**Comment by dascandy** [comment]
The cmakefiles that you moved were those used by Appveyor and Travis to check that the code is not broken... is there a way to reinstate those builds?

File: project/src/CMakeListsTxt.cpp [hunk 1]
```diff
@@ -0,0 +1,227 @@
+#include "../../utils/include/utils.h"
```
Comment (dascandy) [inline project/src/CMakeListsTxt.cpp]: Please don't use relative includes

File: project/src/CMakeListsTxt.cpp [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+            if(comp.second.type == "library")
+            {
+                target_type = "add_library(";
+                target_modifier = " STATIC";
+
+                if(comp.second.isHeaderOnly())
+                {
+                    target_modifier = " INTERFACE";
+                    publicAccess = "INTERFACE";
+                    privateAccess = "INTERFACE"; // We probably should just abort here
```
Comment (dascandy) [inline project/src/CMakeListsTxt.cpp]: Can you change the comment to indicate why this is?

File: toolsets/src/ubuntu.cpp [hunk 3]
```diff
@@ -167,3 +170,33 @@ void UbuntuToolset::CreateCommandsFor(Project &project)
         }
     }
 }
+
+GlobalOptions UbuntuToolset::ParseGeneralOptions(const std::string &options)
```
Comment (dascandy) [inline toolsets/src/ubuntu.cpp]: The one design question I still don't know the answer to is whether the toolset should fix the compile / link options, or whether there'll be one set you have that this code then interpolates to how to tell it to your compiler. You've set it up very well for the second one, but I don't know which of the two it should be. Do you have an opinion & rationale?

File: project/src/CMakeListsTxt.cpp (hunk 2 above)
Comment (jonhanks) [reply project/src/CMakeListsTxt.cpp]: I'll clarify the comment.  That was in regards to a header only library having a private dependency.  It just doesn't make sense.

**Comment by dascandy** [comment]
Integrated last master so I can merge this in too. Thanks!

**Comment by dascandy** [comment]
Again, thanks a ton @jonhanks !

**Pull request merged**